action/protocol/staking/candidate_statemanager.go

```diff
@@ -173,3 +180,129 @@ func (csm *candSM) Commit() error {
 	// write update view back to state factory
 	return csm.WriteView(protocolID, ConvertToViewData(csm))
 }
+
+func (csm *candSM) updateBucket(index uint64, bucket *VoteBucket) error {
```

Comment:
think these private methods are to be added to CandidateStateManager interface
so protocol.StateManager can be removed

Reply:
according to  #2142 description

Reply:
also, latest code has changed quite a bit, pls rebase the PR